### PR TITLE
Add a common `Simplificator` in the `Traverser`

### PR DESCRIPTION
**Pull request by sergeypospelov**
# Description

Refactor: 
 - Removed `runBlockingWithCancellationPredicate` in our tests pipeline
 - Moved `ExecutionState` to `org.utbot.engine.state` package. Moved out succeeding operations for `ExecutionState` to `ExecutionStateUpdates.kt`
 - Some minor refactorings

Add:
- A common `Simplifer` which holds a cache for a whole `Traverser`.
- A `MemoryUpdateSimplifier` which simplifies `MemoryUpdate`s.

Problems:
- Also tried to add simplified local memory updates, but didn't succeed. `SetsTest::testSetContainsInteger` was continuously failing. Now they aren't being simplified explicitly, but simplified implicitly with `Traverser::resolve(Value)` function.

## Type of Change

- Refactoring (typos and non-functional changes) 

# How Has This Been Tested?

## Automated Testing

Tests are passing.

## Manual Scenario 

I checked our engine on synthetic test with ~200 groups of array store and array select expressions (see code), and now it's 50 times faster.
```java
public int longMethod(int x, int y) {
    int[] arr = new int[201];
    int cur = 1;

    arr[cur] = arr[cur - 1] + x - y;
    cur++;

    arr[cur] = arr[cur - 1] + x - y;
    cur++;

    // ^^^^
    // 200 times
   
    if (arr[cur - 1] == 400) {
        return 1;
    }
    return 2;
}
```

# Checklist:

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] No new warnings
- [x] All tests pass locally with my changes

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/engine/pc/Query.kt [hunk 1]
```diff
@@ -99,10 +99,10 @@ data class Query(
     private val gts: PersistentMap<UtExpression, Long> = persistentHashMapOf()
 ) : BaseQuery(hard, soft, assumptions, status, lastAdded) {
 
-    val rewriter: RewritingVisitor
-        get() = RewritingVisitor(eqs, lts, gts)
+    val rewriter: Simplificator
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/pc/Query.kt:102]: Should we rename it?

File: utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+typealias ClassIdToClearStaticsType = ClassId?
+typealias InstanceFieldReadsType = PersistentSet<InstanceFieldReadOperation>
+typealias SpeculativelyNotNullAddressesType = PersistentList<UtAddrExpression>
+typealias SymbolicEnumValuesType = PersistentList<ObjectValue>
+
+context(Simplificator)
+class MemoryUpdateSimplificator : CachingSimplificatorAdapter<MemoryUpdate>() {
+    override fun simplifyImpl(expression: MemoryUpdate): MemoryUpdate {
+        val stores = simplifyStores(expression.stores)
+        val touchedChunkDescriptors = simplifyTocuhedChunkDescriptors(expression.touchedChunkDescriptors)
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt:46]: typo, simplifyTo_cuh_edChunkDescriptors

File: utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+typealias VisitedValuesType = PersistentList<UtAddrExpression>
+typealias TouchedAddressesType = PersistentList<UtAddrExpression>
+typealias ClassIdToClearStaticsType = ClassId?
+typealias InstanceFieldReadsType = PersistentSet<InstanceFieldReadOperation>
+typealias SpeculativelyNotNullAddressesType = PersistentList<UtAddrExpression>
+typealias SymbolicEnumValuesType = PersistentList<ObjectValue>
+
+context(Simplificator)
+class MemoryUpdateSimplificator : CachingSimplificatorAdapter<MemoryUpdate>() {
+    override fun simplifyImpl(expression: MemoryUpdate): MemoryUpdate {
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt:44]: Maybe better to use `with (expression)`?

File: utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    private fun simplifyStaticFieldsUpdates(staticFieldsUpdates: StaticFieldsUpdatesType): StaticFieldsUpdatesType =
+        staticFieldsUpdates.mutate { prevUpdates ->
+            prevUpdates.replaceAll { staticFieldMemoryUpdateInfo(it) }
+        }
+
+    private fun simplifyMeaningfulStaticFields(meaningfulStaticFields: MeaningfulStaticFieldsType): MeaningfulStaticFieldsType =
+        meaningfulStaticFields
+
+
+    private fun simplifyAddrToArrayType(addrToArrayType: AddrToArrayTypeType): AddrToArrayTypeType =
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt:128]: `AddrToArrayTypeType` a bit odd name

File: utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        addrToArrayType
+            .mapKeys { (k, _) -> k.accept(this@Simplificator) as UtAddrExpression }
+            .toPersistentMap()
+
+
+    private fun simplifyAddToMockInfo(addrToMockInfo: AddrToMockInfoType): AddrToMockInfoType =
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt:134]: `add` or `addr`?

File: utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
+package org.utbot.engine.state
+
+import kotlinx.collections.immutable.PersistentMap
+import kotlinx.collections.immutable.persistentHashMapOf
+import org.utbot.engine.LocalMemoryUpdate
+import org.utbot.engine.LocalVariable
+import org.utbot.engine.Parameter
+import org.utbot.engine.SymbolicValue
+//import org.utbot.engine.symbolic.simplificators.LocalMemoryUpdateSimplificator
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt:9]: import

File: utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    val doesntThrow: Boolean = false,
+    val method: SootMethod,
+) {
+    fun update(memoryUpdate: LocalMemoryUpdate, doesntThrow: Boolean = this.doesntThrow) =
+        this.copy(localVariableMemory = localVariableMemory.update(memoryUpdate), doesntThrow = doesntThrow)
+}
+
+/**
+ * Represents a memory associated with a certain method call. For now consists only of local variables mapping.
+ * TODO: think on other fields later
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt:35]: Do we have a task for it? And what fields might be required here?

File: utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    fun local(variable: LocalVariable): SymbolicValue? = locals[variable]
+
+    val localValues: Set<SymbolicValue>
+        get() = locals.values.toSet()
+}
+//
+//context(LocalMemoryUpdateSimplificator)
+//fun updateSimplified(localVariableMemory: LocalVariableMemory, update: LocalMemoryUpdate): LocalVariableMemory {
+//    return localVariableMemory.update(simplify(update))
+//}
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt:58]: Commented code

**Review by CaelmBleidd — CHANGES_REQUESTED** [review]

File: utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt (hunk 4 above)
Comment (sergeypospelov) [reply utbot-framework/src/main/kotlin/org/utbot/engine/simplificators/MemoryUpdateSimplificator.kt:128]: I agree, but I'd like it to be unified with other typealiases.

**Review by CaelmBleidd — APPROVED** [review]

File: utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt (hunk 7 above)
Comment (sergeypospelov) [reply utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt]: I've just moved this class from `ExecutionState.kt` and yes, there are connected issues: #339, #340

Comment (sergeypospelov) [reply utbot-framework/src/main/kotlin/org/utbot/engine/state/ExecutionStackElement.kt]: Added a new comment anyway

**Pull request merged**